pkg/auth/requests/service_account.go

```
@@ -0,0 +1,161 @@
+		return &user.DefaultInfo{}, false, nil
+	}
+
+	if !settings.Enabled {
+		logrus.Warnf("rejecting downstream proxy request for %s, current setting is enabled: %v", req.URL.Path, settings.Enabled)
+		return &user.DefaultInfo{}, false, nil
+	}
+
+	// See if the token is a JWT.
+	rawToken := tokens.GetTokenAuthFromRequest(req)
+
+	jwtParser := jwtv4.Parser{}
+	claims := jwtv4.RegisteredClaims{}
+	_, _, err = jwtParser.ParseUnverified(rawToken, &claims)
+	if err != nil {
+		logrus.Debugf("saauth: error parsing JWT: %v", err)
+		return info, false, err
+	}
+
+	if !strings.HasPrefix(claims.Subject, serviceAccountSubjectPrefix) {
+		logrus.Debugf("saauth: JWT sub is not a service account: %v", err)
+		return info, false, nil
+	}
+
+	if isTokenExpired(claims.ExpiresAt) {
+		logrus.Debugf("saauth: Service Account JWT is expired. Expiration time was: %v", claims.ExpiresAt)
+		return info, false, nil
+	}
+
+	cluster, err := t.clusterLister.Get("", clusterID)
+	if err != nil {
+		logrus.Debugf("saauth: error getting cluster: %v", err)
+		return info, false, err
+	}
+
+	// Get rest config for the cluster and instantiate an authentication client.
+	kubeConfig, err := t.restConfigGetter(cluster, t.scaledContext, t.secretLister)
+	if kubeConfig == nil || err != nil {
+		return info, false, err
+	}
+
+	authClient, err := t.authClientCreator(kubeConfig)
+	if err != nil {
+		logrus.Debugf("saauth: error creating authentication client: %v", err)
+		return info, false, err
+	}
+
+	tokenReview := &v1.TokenReview{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "authentication.k8s.io/v1",
+			Kind:       "TokenReview",
+		},
+		Spec: v1.TokenReviewSpec{
+			Token: rawToken,
+		},
+	}
+
+	// Make the token review request to the downstream cluster.
+	tokenReview, err = authClient.TokenReviews().Create(req.Context(), tokenReview, metav1.CreateOptions{})
+	if err != nil {
+		logrus.Debugf("saauth: error creating a tokenreview request: %v", err)
+		return info, false, nil
```

I think it would be good to return an error in this case since the end user supplied sufficient auth information, but we weren't able to validate it.

---

The thought here was to avoid returning an error which might prevent other authenticators in the chain from being attempted.

---

I think we should also return the error here as we do when the downtream client can't be created. More info [here](https://github.com/rancher/rancher/pull/43147/files#r1530002656)